### ISSUE TITLE
Segmentation Fault when Fuzzing CPP Example

### ISSUE THREAD

**Issue by JannisKirschner** [xiph/flac#162]
- GDB Output:
```
gdb-peda$ c
Continuing.
sample rate    : 11025 Hz
channels       : 2
bits per sample: 16
total samples  : 7078

Program received signal SIGSEGV, Segmentation fault.
[----------------------------------registers-----------------------------------]
RAX: 0x0
RBX: 0x0
RCX: 0x149d7b0 --> 0x0
RDX: 0x616000000080 --> 0xbebebebefbad2c84
RSI: 0x616000000080 --> 0xbebebebefbad2c84
RDI: 0x0
RBP: 0x616000000080 --> 0xbebebebefbad2c84
RSP: 0x7fffffffddd0 --> 0x1000
RIP: 0x5181e3 (<OurDecoder::write_callback(FLAC__Frame const*, int const* const*)+659>: movzx  ebx,WORD PTR [rdi])
R8 : 0x0
R9 : 0x621000001500 --> 0x6ebc46464952
R10: 0x0
R11: 0x7ffff68f87e0 (<fputc>:   test   BYTE PTR [rsi+0x74],0x80)
R12: 0x0
R13: 0xfffffffffffffffc
R14: 0xffffffffc56 --> 0x0
R15: 0x0
EFLAGS: 0x10246 (carry PARITY adjust ZERO sign trap INTERRUPT direction overflow)
[-------------------------------------code-------------------------------------]
   0x5181d4 <OurDecoder::write_callback(FLAC__Frame const*, int const* const*)+644>:    movzx  eax,BYTE PTR [rax+0x7fff8000]
   0x5181db <OurDecoder::write_callback(FLAC__Frame const*, int const* const*)+651>:    test   al,al
   0x5181dd <OurDecoder::write_callback(FLAC__Frame const*, int const* const*)+653>:
    jne    0x51829b <OurDecoder::write_callback(FLAC__Frame const*, int const* const*)+843>:        jne    0x51829b <OurDecoder::write_callback(FL                         AC__Frame const*, int const* const*)+843>
=> 0x5181e3 <OurDecoder::write_callback(FLAC__Frame const*, int const* const*)+659>:    movzx  ebx,WORD PTR [rdi]
   0x5181e6 <OurDecoder::write_callback(FLAC__Frame const*, int const* const*)+662>:    mov    edi,ebx
   0x5181e8 <OurDecoder::write_callback(FLAC__Frame const*, int const* const*)+664>:    mov    rsi,rbp
   0x5181eb <OurDecoder::write_callback(FLAC__Frame const*, int const* const*)+667>:    call   0x41ac30 <fputc@plt>
   0x5181f0 <OurDecoder::write_callback(FLAC__Frame const*, int const* const*)+672>:    cmp    eax,0xffffffff
[------------------------------------stack-------------------------------------]
0000| 0x7fffffffddd0 --> 0x1000
0008| 0x7fffffffddd8 --> 0xc460000002f --> 0x0
0016| 0x7fffffffdde0 --> 0xc46000000ca --> 0x0
0024| 0x7fffffffdde8 --> 0x7fffffffe2b0 --> 0x616000000080 --> 0xbebebebefbad2c84
0032| 0x7fffffffddf0 --> 0x623000000650 --> 0x2b1100001000
0040| 0x7fffffffddf8 --> 0x623000000178 --> 0x629000000210 --> 0x300000000 --> 0x0
0048| 0x7fffffffde00 --> 0x623000000180 --> 0x0
0056| 0x7fffffffde08 --> 0x7fffffffdfe0 --> 0xc0c00000005 --> 0x0
[------------------------------------------------------------------------------] blue
Legend: code, data, rodata, value
Stopped reason: SIGSEGV
0x00000000005181e3 in OurDecoder::write_callback (this=0x7fffffffe2b0, frame=0x623000000650, buffer=0x623000000178)
    at /home/jannis/AFL/c/flac/flac/examples/cpp/decode/file/main.cpp:161
161                             !write_little_endian_int16(f, (FLAC__int16)buffer[1][i])     /* right channel */
gdb-peda$ bt
#0  0x00000000005181e3 in OurDecoder::write_callback (this=0x7fffffffe2b0, frame=0x623000000650, buffer=0x623000000178)
    at /home/jannis/AFL/c/flac/flac/examples/cpp/decode/file/main.cpp:161
#1  0x00000000005313f9 in write_audio_frame_to_client_ (decoder=0x602000000010, frame=0x623000000650, buffer=0x623000000178)
    at /home/jannis/AFL/c/flac/flac/src/libFLAC/stream_decoder.c:2982
#2  read_frame_ (decoder=0x602000000010, got_a_frame=0x7fffffffe1a0, do_full_decode=<optimized out>)
    at /home/jannis/AFL/c/flac/flac/src/libFLAC/stream_decoder.c:2153
#3  0x0000000000533dde in FLAC__stream_decoder_process_until_end_of_stream (decoder=0x602000000010)
    at /home/jannis/AFL/c/flac/flac/src/libFLAC/stream_decoder.c:1089
#4  0x0000000000519ee1 in FLAC::Decoder::Stream::process_until_end_of_stream (this=<optimized out>)
    at /home/jannis/AFL/c/flac/flac/src/libFLAC++/stream_decoder.cpp:218
#5  0x0000000000517be9 in main (argc=<optimized out>, argc@entry=0x3, argv=argv@entry=0x7fffffffe448)
    at /home/jannis/AFL/c/flac/flac/examples/cpp/decode/file/main.cpp:112
#6  0x00007ffff6892b97 in __libc_start_main (main=0x5177b0 <main(int, char**)>, argc=0x3, argv=0x7fffffffe448, init=<optimized out>,
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe438) at ../csu/libc-start.c:310
#7  0x000000000041af5a in _start ()
```

- ASAN Output:
```
jannis@Testbox:~/AFL/c/flac/examples/cpp/decode/file/findings/crashes$ ../../decode_file_cxx id\:00000
id:000000,sig:06,src:000000,op:flip1,pos:20  id:000001,sig:06,src:000019,op:flip1,pos:20
jannis@Testbox:~/AFL/c/flac/examples/cpp/decode/file/findings/crashes$ ../../decode_file_cxx id\:000000\,sig\:06\,src\:000000\,op\:flip1\,pos\:20 /dev/null
sample rate    : 11025 Hz
channels       : 2
bits per sample: 16
total samples  : 7078
AddressSanitizer:DEADLYSIGNAL
=================================================================
==56442==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x0000005181e3 bp 0x616000000080 sp 0x7ffdc44a3330 T0)
==56442==The signal is caused by a READ memory access.
==56442==Hint: address points to the zero page.
    #0 0x5181e2 in OurDecoder::write_callback(FLAC__Frame const*, int const* const*) /home/jannis/AFL/c/flac/flac/examples/cpp/decode/file/main.cpp:161:47
    #1 0x5313f8 in write_audio_frame_to_client_ /home/jannis/AFL/c/flac/flac/src/libFLAC/stream_decoder.c:2982:10
    #2 0x5313f8 in read_frame_ /home/jannis/AFL/c/flac/flac/src/libFLAC/stream_decoder.c:2153
    #3 0x533ddd in FLAC__stream_decoder_process_until_end_of_stream /home/jannis/AFL/c/flac/flac/src/libFLAC/stream_decoder.c:1089:9
    #4 0x519ee0 in FLAC::Decoder::Stream::process_until_end_of_stream() /home/jannis/AFL/c/flac/flac/src/libFLAC++/stream_decoder.cpp:218:29
    #5 0x517be8 in main /home/jannis/AFL/c/flac/flac/examples/cpp/decode/file/main.cpp:112:16
    #6 0x7fb5f4eaeb96 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310
    #7 0x41af59 in _start (/home/jannis/AFL/c/flac/examples/cpp/decode/file/decode_file_cxx+0x41af59)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /home/jannis/AFL/c/flac/flac/examples/cpp/decode/file/main.cpp:161:47 in OurDecoder::write_callback(FLAC__Frame const*, int const* const*)
==56442==ABORTING
```
[crashes.zip](https://github.com/xiph/flac/files/3846778/crashes.zip)

**Comment by neheb** [comment]
Wonder if it's related to the clang-tidy fixes.

**Comment by erikd** [comment]
@JannisKirschner What version of git hash is this?

**Comment by JannisKirschner** [comment]
@erikd should be the current version. 
The latest commit hash is 2907d4921cc4e46751c18478f5cbe1d60ce50cac.
Build on Ubuntu 18.04.2 LTS.

**Comment by ktmf01** [comment]
It turns out this very same problem has been fixed previously in the C decoder example: 93846ee

Original bug report here: https://sourceforge.net/p/flac/bugs/418/
